src/Vendr.PaymentProviders.Worldpay/WorldpayPaymentProvider.cs

```diff
@@ -1,38 +1,139 @@
         {
-            return new PaymentFormResult()
+            try
+            {
+                if (settings.VerboseLogging)
+                {
+                    _logger.Info<WorldpayPaymentProvider>($"GenerateForm method called for cart {order.CartNumber}");
+                }
+
+                var url = settings.Mode.ToLower() == "live" ? settings.LiveUrl : settings.TestUrl;
+                var form = new PaymentForm(url, FormMethod.Post);
+
+                settings.InstallId.MustNotBeNull("settings.InstallId");
+                settings.TestModeNumber.MustNotBeNull("settings.TestModeNumber");
+                settings.AuthMode.MustNotBeNull("settings.AuthMode");
+                settings.OrderPropertyBillingFirstName.MustNotBeNull("settings.OrderPropertyBillingFirstName");
+                settings.OrderPropertyBillingLastName.MustNotBeNull("settings.OrderPropertyBillingLastName");
+
+                var firstname = order.Properties[settings.OrderPropertyBillingFirstName];
+                var surname = order.Properties[settings.OrderPropertyBillingLastName];
+                var address1 = order.Properties[settings.OrderPropertyBillingAddress1] ?? string.Empty;
+                var city = order.Properties[settings.OrderPropertyBillingCity] ?? string.Empty;
+                var postcode = order.Properties[settings.OrderPropertyBillingPostcode] ?? string.Empty;
+                var billingCountry = Vendr.Services.CountryService.GetCountry(order.PaymentInfo.CountryId.Value);
+                var billingCountryCode = billingCountry.Code.ToUpperInvariant();
+                var amount = order.TotalPrice.Value.WithTax.ToString("0.00", CultureInfo.InvariantCulture);
+                var currency = Vendr.Services.CurrencyService.GetCurrency(order.CurrencyId);
+                var currencyCode = currency.Code.ToUpperInvariant();
+
+                // Ensure billing country has valid ISO 3166 code
+                var iso3166Countries = Vendr.Services.CountryService.GetIso3166CountryRegions();
+                if (iso3166Countries.All(x => x.Code != billingCountryCode))
+                {
+                    throw new Exception("Country must be a valid ISO 3166 billing country code: " + billingCountry.Name);
+                }
+
+                // Ensure currency has valid ISO 4217 code
+                if (!Iso4217.CurrencyCodes.ContainsKey(currencyCode))
+                {
+                    throw new Exception("Currency must be a valid ISO 4217 currency code: " + currency.Name);
+                }
+
+                var orderDetails = new Dictionary<string, string>
+                {
+                    { "instId", settings.InstallId },
+                    { "testMode", settings.TestModeNumber },
+                    { "authMode", settings.AuthMode },
+                    { "cartId", order.CartNumber },
```

Comment:
I'd probably use `order.OrderNumber` instead of order.CartNumber. Vendr is slightly different to TC in that it generates an order number prior to going to the payment gateway.